src/private.rs

```diff
@@ -72,100 +61,118 @@ impl RedactFlags {
-    let mut redacted = String::new();
-
-    let to_redactable_string = || match this {
-        RedactionTarget::Debug { this, alternate: false } => format!("{:?}", this),
-        RedactionTarget::Debug { this, alternate: true } => format!("{:#?}", this),
-        RedactionTarget::Display(this) => this.to_string(),
-    };
-
+impl RedactionTarget<'_> {
+    /// Pass through directly to the formatter.
     #[cfg(feature = "toggle")]
-    if crate::toggle::get_redaction_behavior().is_plaintext() {
-        return DisplayDebug(to_redactable_string());
+    pub(crate) fn passthrough(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            RedactionTarget::Debug { this, .. } => std::fmt::Debug::fmt(this, fmt),
+            RedactionTarget::Display(this) => std::fmt::Display::fmt(this, fmt),
+        }
     }
+}
+impl ToString for RedactionTarget<'_> {
+    fn to_string(&self) -> String {
+        match self {
+            RedactionTarget::Debug { this, alternate: false } => format!("{:?}", this),
+            RedactionTarget::Debug { this, alternate: true } => format!("{:#?}", this),
+            RedactionTarget::Display(this) => this.to_string(),
+        }
+    }
+}
 
-    (|| {
-        if flags.fixed > 0 {
-            flags.redact_fixed(flags.fixed as usize, &mut redacted);
-            return;
+pub struct RedactionFormatter<'a> {
+    pub this: RedactionTarget<'a>,
+    pub flags: RedactFlags,
+    pub specialization: Option<RedactSpecialization>,
+}
+impl std::fmt::Debug for RedactionFormatter<'_> {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        #[cfg(feature = "toggle")]
+        if crate::toggle::get_redaction_behavior().is_plaintext() {
+            return self.this.passthrough(fmt);
         }
 
-        let redactable_string = to_redactable_string();
-
-        redacted.reserve(redactable_string.len());
-
-        // Specialize for Option<T>
-        if flags.is_option {
-            if redactable_string == "None" {
-                // We don't need to do any redacting
-                // https://prima.slack.com/archives/C03URH9N43U/p1661423554871499
-            } else if let Some(inner) = redactable_string
-                .strip_prefix("Some(")
-                .and_then(|inner| inner.strip_suffix(')'))
-            {
-                redacted.push_str("Some(");
-                flags.redact_partial(inner, &mut redacted);
-                redacted.push(')');
-            } else {
-                // This should never happen, but just in case...
-                flags.redact_full(&redactable_string, &mut redacted);
+        if self.flags.fixed > 0 {
+            return RedactFlags::redact_fixed(fmt, self.flags.fixed as usize, self.flags.redact_char);
+        }
+
+        let redactable_string = self.this.to_string();
+
+        #[allow(clippy::single_match)]
+        match self.specialization {
```

Comment:
why match instead of just if let?

Reply:
We may introduce new specialisations in future, so it would look kind of weird to have an `if let`, which kinda implies there will only ever be one kind of specialisation. Unfortunately `#[non_exhaustive]` has no effect too in the same crate, so that doesn't fix the clippy warning :(

I'll slap on a `#[non_exhaustive]` anyway

Reply:
Moved to https://github.com/primait/veil/pull/60 where it belongs